### PR TITLE
feat: add json parser to build attribute in duktape

### PR DESCRIPTION
**Pull request by sheny1xuan**
Signed-off-by: stonex <1479765922@qq.com>

### Description
+ Link to this [issue](https://github.com/casbin/casbin-cpp/issues/165).
+ Add a json parser to build duktape object.
+ Test `double`, `int`, `bool`, `string` and `object`.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@xcaptain

File: tests/model_enforcer_test.cpp [hunk 1]
```diff
@@ -504,6 +504,25 @@ TEST(TestModelEnforcer, TestRBACModelWithPattern) {
...
+
+TEST(TestModelEnforcer, TestABACModelWithJson) {
+    using json = nlohmann::json;
+
+    json obj = {
+            {"Owner", "alice"},
+        };
+    auto obj_ptr = std::make_shared<json>(obj);
+
+    casbin::DataMap MapParams = {{"sub", "alice"}, {"obj", obj_ptr}, {"act", "write"}};
```
Comment (EmperorYP7) [inline tests/model_enforcer_test.cpp:516]:
```suggestion
    casbin::DataMap mapParams = {{"sub", "alice"}, {"obj", obj_ptr}, {"act", "write"}};
```

File: tests/model_enforcer_test.cpp [hunk 2]
```diff
@@ -504,6 +504,25 @@ TEST(TestModelEnforcer, TestRBACModelWithPattern) {
...
+    casbin::DataList ListParams = {"alice", obj_ptr, "write"};
```
Comment (EmperorYP7) [inline tests/model_enforcer_test.cpp:517]:
```suggestion
    casbin::DataList listParams = {"alice", obj_ptr, "write"}; 
```

File: tests/model_enforcer_test.cpp [hunk 3]
```diff
@@ -504,6 +504,25 @@ TEST(TestModelEnforcer, TestRBACModelWithPattern) {
...
+    using json = nlohmann::json;
+
+    json obj = {
+            {"Owner", "alice"},
+        };
+    auto obj_ptr = std::make_shared<json>(obj);
+
+    casbin::DataMap MapParams = {{"sub", "alice"}, {"obj", obj_ptr}, {"act", "write"}};
+    casbin::DataList ListParams = {"alice", obj_ptr, "write"}; 
+    casbin::DataVector VectorParams = {"alice", obj_ptr, "write"};
```
Comment (EmperorYP7) [inline tests/model_enforcer_test.cpp:518]:
```suggestion
    casbin::DataVector vectorParams = {"alice", obj_ptr, "write"}; 
```

File: tests/model_enforcer_test.cpp [hunk 4]
```diff
@@ -504,6 +504,25 @@ TEST(TestModelEnforcer, TestRBACModelWithPattern) {
...
+
+    casbin::DataMap MapParams = {{"sub", "alice"}, {"obj", obj_ptr}, {"act", "write"}};
+    casbin::DataList ListParams = {"alice", obj_ptr, "write"}; 
+    casbin::DataVector VectorParams = {"alice", obj_ptr, "write"}; 
+
+    casbin::Enforcer e(abac_model_path);
+
+    ASSERT_TRUE(e.Enforce(MapParams));
+    ASSERT_TRUE(e.Enforce(ListParams));
+    ASSERT_TRUE(e.Enforce(VectorParams));
```
Comment (EmperorYP7) [inline tests/model_enforcer_test.cpp:524]: likewise

File: casbin/model/scope_config.cpp [hunk 5]
```diff
@@ -194,6 +194,38 @@ void PushObjectPropToObject(Scope scope, std::string obj, std::string identifier
...
+        } else if (value.is_number_float()) {
+            PushDoublePropToObject(scope, j_name, value, key);
+        } else if (value.is_number_integer()) {
+            PushIntPropToObject(scope, j_name, value, key);
+        } else if (value.is_string()) {
+            PushStringPropToObject(scope, j_name, value, key);
+        } else if (value.is_boolean()) {
+            PushBooleanPropToObject(scope, j_name, value, key);
+        } else {
+            throw IllegalArgumentException("Unsupport json value type");
```
Comment (EmperorYP7) [inline casbin/model/scope_config.cpp:224]:
```suggestion
            throw IllegalArgumentException("Unsupported json value type");
```

**Review by EmperorYP7 — APPROVED** [review]
LGTM! Well done @sheny1xuan 🙌🏻 However, the variable naming should be consistent throughout the code. (pascalCase)

**Comment by EmperorYP7** [comment]
I was wondering if we actually need the `casbin::ABACData` workflow now. Since the JSON parser can handle single depth attributes as well. What do you think @sheny1xuan?

**Comment by hsluoyz** [comment]
@EmperorYP7 then we can remove it I think

File: tests/model_enforcer_test.cpp (hunk 1 above)
Comment (sheny1xuan) [reply tests/model_enforcer_test.cpp]: I've done it, thanks.

**Comment by sheny1xuan** [comment]
I think so, but I think we should make JSON parse comprehensive support the old work that `casbin::ABACData` do. Then we can remove it. @EmperorYP7

**Comment by hsluoyz** [comment]
@sheny1xuan good point!

**Pull request merged**